samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -497,20 +546,21 @@ int consume_device_message(void)
 	LOG_DBG("Attempting to transmit enqueued device message");
 
 	struct nrf_cloud_tx_data mqtt_msg = {
-		.data.ptr = msg,
-		.data.len = strlen(msg),
 		.qos = MQTT_QOS_1_AT_LEAST_ONCE,
 		.topic_type = NRF_CLOUD_TOPIC_MESSAGE,
+		.obj = queued_msg
 	};
 
+	/* Send message */
 	ret = nrf_cloud_send(&mqtt_msg);
+
 	if (ret) {
 		LOG_ERR("Transmission of enqueued device message failed, nrf_cloud_send "
 			"gave error: %d. The message will be re-enqueued and tried again "
 			"later.", ret);
 
-		/* Re-enqueue the message for later retry (creating a new copy of it in heap). */
-		send_device_message(msg);
+		/* Re-enqueue the message for later retry */
+		enqueue_device_message(queued_msg, false);
```

Comment:
If `enqueue_device_message` fails, here, the message object will never be freed

Reply:
ah good catch, i guess the return value of `enqueue_device_message` should be checked 🙈